client/driver/executor/executor.go

```diff
@@ -0,0 +1,217 @@
+type ExecCommand struct {
+	Cmd  string
+	Args []string
+}
+
+// ProcessState holds information about the state of
+// a user process
+type ProcessState struct {
+	Pid      int
+	ExitCode int
+	Time     time.Time
+}
+
+// Executor is the interface which allows a driver to launch and supervise
+// a process user wants to run
+type Executor interface {
+	LaunchCmd(command *ExecCommand, ctx *ExecutorContext) (*ProcessState, error)
+	Wait() (*ProcessState, error)
+	ShutDown() error
+	Exit() error
+}
+
+// UniversalExecutor is an implementation of the Executor which launches and
+// supervises processes. In addition to process supervision it provides resource
+// and file system isolation
+type UniversalExecutor struct {
+	cmd exec.Cmd
+	ctx *ExecutorContext
+
+	taskDir       string
+	groups        *cgroupConfig.Cgroup
+	exitState     *ProcessState
+	processExited chan interface{}
+
+	logger *log.Logger
+	lock   sync.Mutex
+}
+
+// NewExecutor returns an Executor
+func NewExecutor(logger *log.Logger) Executor {
+	return &UniversalExecutor{logger: logger, processExited: make(chan interface{})}
+}
+
+// LaunchCmd launches a process and returns it's state. It also configures an
+// applies isolation on certain platforms.
+func (e *UniversalExecutor) LaunchCmd(command *ExecCommand, ctx *ExecutorContext) (*ProcessState, error) {
+	e.logger.Printf("[INFO] executor: launching command %v", command.Cmd)
+
+	e.ctx = ctx
+
+	// configuring the task dir
+	if err := e.configureTaskDir(); err != nil {
+		return nil, err
+	}
+
+	// confiuguring the chroot
+	if err := e.configureIsolation(); err != nil {
+		return nil, err
+	}
+
+	// setting the user of the process
+	if e.ctx.UnprivilegedUser {
+		if err := e.runAs("nobody"); err != nil {
+			return nil, err
+		}
+	}
+
+	// configuring log rotate
+	stdoPath := filepath.Join(e.taskDir, allocdir.TaskLocal, fmt.Sprintf("%v.stdout", ctx.TaskName))
+	stdo, err := os.OpenFile(stdoPath, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0666)
+	if err != nil {
+		return nil, err
+	}
+	e.cmd.Stdout = stdo
+
+	stdePath := filepath.Join(e.taskDir, allocdir.TaskLocal, fmt.Sprintf("%v.stderr", ctx.TaskName))
+	stde, err := os.OpenFile(stdePath, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0666)
+	if err != nil {
+		return nil, err
+	}
+	e.cmd.Stderr = stde
+
+	// setting the env, path and args for the command
+	e.cmd.Env = ctx.TaskEnv.EnvList()
+	e.cmd.Path = ctx.TaskEnv.ReplaceEnv(command.Cmd)
+	e.cmd.Args = append([]string{e.cmd.Path}, ctx.TaskEnv.ParseAndReplace(command.Args)...)
+	if filepath.Base(command.Cmd) == command.Cmd {
+		if lp, err := exec.LookPath(command.Cmd); err != nil {
+		} else {
+			e.cmd.Path = lp
+		}
+	}
+
+	// starting the process
+	if err := e.cmd.Start(); err != nil {
+		return nil, fmt.Errorf("error starting command: %v", err)
+	}
+
+	// entering the user process in the cgroup
+	e.applyLimits(e.cmd.Process.Pid)
+	// entering the plugin process in cgroup
+	e.applyLimits(os.Getpid())
```

Comment:
This should happen before you start the user process, otherwise they can escape the cgroups by forking immediately and even escape supervision by fork/execing and then exiting.

It should also be enough to set the plugin in the cgroup as it is the parent of the other PID.